### ISSUE TITLE
JA4 support in haproxy

### ISSUE THREAD

**Issue by awlx** [haproxy/haproxy#2495]
### Your Feature Request

As we already can leverage JA3 fingerprinting in haproxy to do client fingerprinting. 

I would like to see a builtin JA4 hash constant for even better client fingerprinting.

Details about JA4:
https://github.com/FoxIO-LLC/ja4/blob/main/technical_details/README.md

### What are you trying to do?

Client fingerprinting for Botdetection and the like.

### Output of `haproxy -vv`

2.8.7

**Comment by NiceRath** [comment]
Just read about JA4 in the [Cloudflare Blog](https://blog.cloudflare.com/ja4-signals) and [FoxIO Blog](https://blog.foxio.io/ja4%2B-network-fingerprinting)

This feature would be very nice to have. At least in the enterprise edition - as JA3 is currently.

Especially as FoxIO provides a [Fingerprint-DB](https://ja4db.com/) that could be utilized.

**Comment by wtarreau** [comment]
Thanks for the links to the articles, they're super interesting. I'm not aware of anything planned around this yet, but it's probably only a matter of time before someone jumps on it and implements it (maybe within certain technical limitations).

**Comment by NiceRath** [comment]
Yeah. I'll give a LUA implementation a go when I find time to do so.
Though LUA isn't that performant..

**Comment by Tristan971** [comment]
For JA3 it is possible to do it entirely with fetches and converters (in CE) already.

Had a cursory look, and JA4 probably only needs a couple of extra fetches being exposed; the main task would be to go through them 1-by-1 and list the ones not available yet I reckon.

**Comment by superstes** [comment]
Have created a LUA plugin draft: https://github.com/O-X-L/haproxy-ja4
Contributions are welcome!

**Comment by mat1010** [comment]
We are currently struggling getting the tls extension  values for `supported_versions` and `signature_algorithms`. It seem those are not exposed through a variable right now. Would this be worthy of a feature request, or is there a slim way of getting this information within HAProxy / the lua extension?

**Comment by wlallemand** [comment]
Indeed these parameters are not kept in the SSL session so we need to add them in the list of "tune.ssl.capture-buffer-size", I'll take a look at these, thanks.

**Comment by wlallemand** [comment]
@mat1010 I made this [branch](https://github.com/haproxy/haproxy/tree/20240823-fc_supported_versions) containing the `ssl_fc_supported_versions_bin` sample fetch and the `ssl_fc_sigalgs_bin` sample fetch, could you test if this is enough for your script?

**Comment by mat1010** [comment]
@wlallemand Wow, thank you for the quick implementation. `ssl_fc_supported_versions_bin` works exactly like expected `51914-772-771` 👍. The parameter to exclude the GREASE values, as it exists with the other functions, would be nice as well.

Currently `ssl_fc_sigalgs_bin` seems to return an empty table. I will check what's happening here / if my client causes the issue.

**Comment by wlallemand** [comment]
That's how I tested them FYI:

```
frontend fr1
	mode http
	bind *:8443 ssl crt test.pem
	http-request return content-type text/plain status 200 string "test\n"

	http-after-response set-header X-SSL-1 %[ssl_fc_supported_versions_bin,be2hex(' ',2)]
	http-after-response set-header X-SSL-2 %[ssl_fc_sigalgs_bin,be2hex(' ',2)]
```

```
$ curl -I -k https://localhost:8443
HTTP/2 200 
content-length: 12
content-type: text/plain
x-ssl-1: 0304 0303
x-ssl-2: 0403 0503 0603 0807 0808 0809 080A 080B 0804 0805 0806 0401 0501 0601 0303 0301 0302 0402 0502 0602
```

Note that it is extracting the "signature_algorithms" extension, not the "signature_algorithms_cert" extension. https://www.ietf.org/archive/id/draft-ietf-tls-rfc8446bis-03.html#name-signature-algorithms

I'll take a look at the GREASE parameters next week.

**Comment by mat1010** [comment]
Yes - I can confirm `ssl_fc_sigalgs_bin` works as well as expected. There was a bug in my script where I always returned an empty table -.- Thanks a lot again and have a nice weekend.

**Comment by wlallemand** [comment]
Thanks for the feedback, I pushed the final version in the master branch with the GREASE parameter like the other fetches.

**Comment by mat1010** [comment]
Great, thank you. This also works like expected :+1: 

Btw.: If someone is interested, the plugin in https://github.com/O-X-L/haproxy-ja4 now fully works with the latest haproxy master branch.

**Comment by amsnek** [comment]
This is very intriguing! Will the final versions (ssl_fc_supported_versions_bin/ssl_fc_sigalgs_bin) be available in an upcoming 3.0.x release or in 3.1+ ? 
Would it be possible relying only on converters and fetches or does it require (the very interesting) lua scripts? Performance wise I would assume an approach to how JA3 is currently possible in CE would be faster? Are there additional requirements or benefits I am missing regarding the lua scripts? 
Edit: looking through the LUA script with limited understanding, it does provide a lot additional logic and functions. Will certainly be testing this asap.

**Comment by wlallemand** [comment]
> This is very intriguing! Will the final versions (ssl_fc_supported_versions_bin/ssl_fc_sigalgs_bin) be available in an upcoming 3.0.x release or in 3.1+ ?

We don't backport features in stable branches, so this will be available in 3.1.

>  Would it be possible relying only on converters and fetches or does it require (the very interesting) lua scripts? Performance wise I would assume an approach to how JA3 is currently possible in CE would be faster? Are there additional requirements or benefits I am missing regarding the lua scripts?

You won't be able to do it without the script, because values need to be sorted and we don't have anything to do that in the converters yet.

**Comment by superstes** [comment]
The JA3 standard is very simple in comparison. (example: [JA3N LUA](https://github.com/O-X-L/haproxy-ja3n/blob/latest/ja3n.lua) vs [JA4 LUA](https://github.com/O-X-L/haproxy-ja4/blob/latest/ja4.lua))

A basic overview of [what is necessary](https://github.com/FoxIO-LLC/ja4/blob/main/technical_details/JA4.md):
* value mapping (part1+2+3)
* create a padded count of items that is capped at 99 (part 4+5)
* sort values in hex format (part7+8)
* exclude specific extensions (part8)
* sha256 hashed data truncated to 12chars (part7+8)

**Comment by amsnek** [comment]
Thank you for the detailed answers. I am now realising this more, will dig into this subject. Results with JA4 are also far more accurate/granular. The Lua Script works nicely in my test setup as well. 
Looking forward even more to HAProxy 3.1 now :)

**Comment by ZeroHans4** [comment]
Hello 3.1 has been released, but I think this is still missing or did i overlook something?

**Comment by wlallemand** [comment]
@ZeroHans4 `%[ssl_fc_supported_versions_bin]` and  `%[ssl_fc_sigalgs_bin]` fetches were integrated in 3.1, so it's possible to use the ja4 lua project mentioned in this ticket.

**Comment by ZeroHans4** [comment]
@wlallemand I see thanks, are there any plans to integrate the necessary helper functions to generate a JA4T fingerprint?

**Comment by wtarreau** [comment]
Not at this point. The Lua extension above seems to already do the job, can probably even be used with older versions, and is easier to adapt in the field if needed. If there was a valid technical reason for why this is a problem, that option could be reconsidered, but in general we try not to absorb others' work "just because we can".

**Comment by ZeroHans4** [comment]
@wtarreau alright, well, from my understanding the script can generate the JA4 and JA4h fingerprint, but I am looking for OS detection which AFAIK can be only done with JA4T, which is not possible using the LUA above, right?

**Comment by wtarreau** [comment]
I already have no idea about the variants, and that alone doubly justifies to keeping it flexible in Lua for now :-)

**Comment by ZeroHans4** [comment]
I see @superstes, do you think it would be possible to create a version of this LUA which gets JA4T?

**Comment by Tristan971** [comment]
JA4T is (afaict) TCP fingerprinting (https://blog.foxio.io/ja4t-tcp-fingerprinting), and I admit it's quite interesting because you can detect things like VPNs/proxies (or encapsulation in general) by looking at MTU vs MSS for example

At least I've been told of a few specific values in the past that catch some common abusive(/ed) open relays in the past, and when I looked I didn't see a way to get those bits of TCP conn information from existing fetches

**Comment by superstes** [comment]
Yeah - I've looked into it too and have not found the necessary fetches.
As I see it, these are the ones that are available: https://github.com/haproxy/haproxy/blob/v3.1.0/src/tcp_sample.c#L537 | Deeper: [proto_tcp](https://github.com/haproxy/haproxy/blob/v3.1.0/src/proto_tcp.c#L889)
We could possibly use [SPOE](https://www.haproxy.com/blog/extending-haproxy-with-the-stream-processing-offload-engine) but IMHO that would be overkill.
I will look into how this could be done on firewall-level (https://github.com/O-X-L/nftables-ja4t)

**Comment by wtarreau** [comment]
Then we'd rather figure if the missing fetches can reasonably be implemented (by "reasonably" I mean "without hindering the critical paths that everyone else uses". Apparently the MSS can be retrieved at any time so it could be done live in the fetch. Whatever else is missing, I don't know, I never looked at the spec. @superstes if you have somewhere the list of needed info, we can have a look at what's missing to estimate the feasability.

**Comment by superstes** [comment]
> @superstes if you have somewhere the list of needed info...


![JA4T](https://raw.githubusercontent.com/FoxIO-LLC/ja4/refs/heads/main/technical_details/JA4T.png)

**Comment by wtarreau** [comment]
Thanks. I'm seeing ordered tcp options mentioned there, which, to the best of my knowledge, is not accessible from userland and not preserved by the kernel while parsing the SYN. The MSS should be easy to get as we already do it on listeners using getsockopt(). That may explain why the vast majority of the cited components supporting it are in fact network-level devices such as firewalls or DPI devices. But it's probably no big deal to produce only the retrievable info anyway.

**Comment by unixfox** [comment]
Is there any performance penalty of using a lua script compared to a function written in C inside the haproxy core?

**Comment by wtarreau** [comment]
Definitely, but on the other hand, Lua can sometimes be better than C because it's totally interruptible in that it runs in sort of a VM, while most C code wouldn't and could cause huge latencies. For example we had to rewrite the maps purging mechanism for updates because all elements were visited in a loop, and users having huge maps (or ACLs) were frequenly seeing long freezes, which sometimes even managed to trigger the watchdog. So I'd just that that normally (i.e. with an infinite development time budget) you'd do everything in C but when the feature isn't worth spending too much time, Lua is very well suited.

**Comment by ZeroHans4** [comment]
@wtarreau alright sounds good, so would it be realistic to implement the JA4T capabilities into haproxy? Would love to use them.

**Comment by ZeroHans4** [comment]
@NiceRath what do you think about this

**Comment by NiceRath** [comment]
@wtarreau mentioned - the `order of tcp options` seem to not be retrievable.
If that is the case, the complete JA4T cannot be implemented.

**Comment by amsnek** [comment]
Could this in theory be accomplished using pcap (like tcpdump) and/or eBPF support or maybe DPDK? Just thinking out loud here, I guess this would be outside the scope of HAProxy. Would love to use thise as well :)

**Comment by awlx** [comment]
I wonder how we could populate the JA4H fingerprint as well:

![Image](https://github.com/user-attachments/assets/bed4adbd-49bf-4214-81be-e6b68d85ea7f)

**Comment by superstes** [comment]
We'll probably need another Lua script..
I've looked into it a bit but had no time to implement it yet 😕

1. Fetch method
2. Fetch req.ver
3. Fetch req.cook_cnt
4. Check for referer HDR
5. Fetch req.hdrs
6. Check HDR
7. Not sure - maybe req.hdr_names?
8. Maybe req.cook_names/req.cook_val?
9. Maybe req.cook_names + req.cook_val?

Available http fetches for reference:
https://github.com/haproxy/haproxy/blob/master/src/http_fetch.c#L2261

**Comment by superstes** [comment]
Update: have just written a basic LUA-script for JA4H: https://github.com/O-X-L/haproxy-ja4h/blob/latest/ja4h.lua
Feel free to test it and give feedback in the repo.

NOTE: JA4H has some license limitations => https://github.com/FoxIO-LLC/ja4/blob/main/LICENSE

**Comment by awlx** [comment]
Awesome @superstes will test and report back!

**Comment by NiceRath** [comment]
FYI: Seems HAProxy-enterprise has now built-in JA4 🥳

**Comment by WRMSRwasTaken** [comment]
I am running this script for quite a while successfully to combat bots in production, but I've noticed it returning wrong results for QUIC connections on 3.3.1.

Looking closer, the `ssl_fc_protocol_hello_id` fetch is returning nothing for QUIC connections. It seems that this is not the only fetch returning nothing as looking at the hash, quite a lot of fields are set to null / 0.

It was working before, because I definitely was fingerprinting QUIC connections in the past, but I am not sure when it broke.

I am currently on 3.3.1 with OpenSSL 3.6.0 (using native QUIC). Switching to AWS-LC made no difference.

€: It *might* have broken with the switch to the new ncbmbuf implementation in 3.3-dev11? I'll test with 3,3-dev10 to see if it works there the upcoming days.

**Comment by Fijxu** [comment]
> I am running this script for quite a while successfully to combat bots in production, but I've noticed it returning wrong results for QUIC connections on 3.3.1.
> 
> Looking closer, the `ssl_fc_protocol_hello_id` fetch is returning nothing for QUIC connections. It seems that this is not the only fetch returning nothing as looking at the hash, quite a lot of fields are set to null / 0.
> 
> It was working before, because I definitely was fingerprinting QUIC connections in the past, but I am not sure when it broke.
> 
> I am currently on 3.3.1 with OpenSSL 3.6.0 (using native QUIC). Switching to AWS-LC made no difference.
> 
> €: It _might_ have broken with the switch to the new ncbmbuf implementation in 3.3-dev11? I'll test with 3,3-dev10 to see if it works there the upcoming days.

About this issue, I did some git bisecting between  `v3.3-dev14` and  `v3.3-dev13` and I have found that commit https://github.com/haproxy/haproxy/commit/fab7da0fd0a67a9e7286839efca95ecdbf3a059e introduced the issue of `ssl_fc_protocol_hello_id` returning `nil`/`null` for QUIC connections, specially lines `2156-2161` in that commit:

https://github.com/haproxy/haproxy/commit/fab7da0fd0a67a9e7286839efca95ecdbf3a059e#diff-22bd931e83fdea8071b4923e5c37fadb20d99a9df21ebf6e105bc2e8fde68c9bR2156-R2161

For HAProxy 3.3.1, commenting lines `2164-2165` (https://git.haproxy.org/?p=haproxy-3.3.git;a=blob;f=src/ssl_sock.c;h=6f3c88d1d7338fb43edf983f61853089f50168a1;hb=refs/tags/v3.3.1#l2164 there is no `v3.3.1` tag in Github) fixes this issue, although that is more like a workaround instead of a proper fix

**Comment by a-denoyelle** [comment]
@Fijxu I just stumbled on a similar issue as SSL keylog is broken for QUIC since 3.3. This is also related to the commit you mentionned.

**Comment by superstes** [comment]
It might be better to create a dedicated bug-issue for this problem as it is not only affecting the JA4-fingerprinting logic (:

**Comment by a-denoyelle** [comment]
We could have open a dedicated issue but I already merged the fix in master and 3.3 repository. The next release should behave as expected.

**Comment by Fijxu** [comment]
> > I am running this script for quite a while successfully to combat bots in production, but I've noticed it returning wrong results for QUIC connections on 3.3.1.
> > Looking closer, the `ssl_fc_protocol_hello_id` fetch is returning nothing for QUIC connections. It seems that this is not the only fetch returning nothing as looking at the hash, quite a lot of fields are set to null / 0.
> > It was working before, because I definitely was fingerprinting QUIC connections in the past, but I am not sure when it broke.
> > I am currently on 3.3.1 with OpenSSL 3.6.0 (using native QUIC). Switching to AWS-LC made no difference.
> > €: It _might_ have broken with the switch to the new ncbmbuf implementation in 3.3-dev11? I'll test with 3,3-dev10 to see if it works there the upcoming days.
> 
> About this issue, I did some git bisecting between `v3.3-dev14` and `v3.3-dev13` and I have found that commit [fab7da0](https://github.com/haproxy/haproxy/commit/fab7da0fd0a67a9e7286839efca95ecdbf3a059e) introduced the issue of `ssl_fc_protocol_hello_id` returning `nil`/`null` for QUIC connections, specially lines `2156-2161` in that commit:
> 
> [fab7da0#diff-22bd931e83fdea8071b4923e5c37fadb20d99a9df21ebf6e105bc2e8fde68c9bR2156-R2161](https://github.com/haproxy/haproxy/commit/fab7da0fd0a67a9e7286839efca95ecdbf3a059e#diff-22bd931e83fdea8071b4923e5c37fadb20d99a9df21ebf6e105bc2e8fde68c9bR2156-R2161)
> 
> For HAProxy 3.3.1, commenting lines `2164-2165` (https://git.haproxy.org/?p=haproxy-3.3.git;a=blob;f=src/ssl_sock.c;h=6f3c88d1d7338fb43edf983f61853089f50168a1;hb=refs/tags/v3.3.1#l2164 there is no `v3.3.1` tag in Github) fixes this issue, although that is more like a workaround instead of a proper fix

I can confirm this issue has been solved with https://github.com/haproxy/haproxy/commit/869a997a68a23ef235aaea16bc80dfb1801ec7ef (https://git.haproxy.org/?p=haproxy-3.3.git;a=commitdiff;h=876b97c29). Thanks!

**Comment by wlallemand** [comment]
I'm closing this, as we already provide sample fetches allowing to generate a JA4 fingerprint.

Regarding JA4+ (JA4S, JA4L, JA4H, JA4X, JA4SSH) this probably won't be directly implemented since FoxIO seems to patent them. https://github.com/FoxIO-LLC/ja4/blob/main/License%20FAQ.md

**Comment by lukastribus** [comment]
FYI Luca Deri (ntop) just released a open, patent free successor, see "Introducing JA5: A Stable, Collision-Resistant Successor to JA4":

https://www.linkedin.com/pulse/introducing-ja5-stable-collision-resistant-successor-ja4-luca-deri-9zbre/

It's about to get merged into Wireshark:
https://gitlab.com/wireshark/wireshark/-/merge_requests/26589

I don't know enough about either JA4 or JA5 to make recommendations or requests, just stating it for the record.

**Comment by wtarreau** [comment]
Pretty interesting, thanks Lukas for the link! It seems to address real shortcomings that were plaguing JA4 for TLS from what is explained in the article. Pretty happy that we didn't waste anyone's energy implementing nor deploying JA4 in the end ;-)

**Comment by amsnek** [comment]
Does this mean HAProxy might consider implementing JA5 / fetches in a future release that would allow using it without lua scripts?
